File: apps/chat/libs/chat-storage/src/main/java/com/example/chat/storage/entity/ChatChannelEntity.java
package com.example.chat.storage.entity;

import java.time.Instant;

import com.example.chat.domain.channel.ChannelType;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

/**
 * 채널 JPA Entity
 */
@Entity
@Table(name = "chat_channels", indexes = {
        @Index(name = "idx_chat_channel_owner", columnList = "owner_id"),
        @Index(name = "idx_chat_channel_type", columnList = "channel_type")
})
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Builder
public class ChatChannelEntity {
    @Id
    @Column(name = "id", length = 36, nullable = false)
    private String id;

    @Column(name = "name", nullable = false, length = 100)
    private String name;

    @Column(name = "description", length = 500)
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "channel_type", nullable = false, length = 20)
    private ChannelType channelType;

    @Column(name = "owner_id", nullable = false, length = 36)
    private String ownerId;

    @Column(name = "active", nullable = false)
    @Builder.Default
    private boolean active = true;

    @Column(name = "created_at", nullable = false)
    @Builder.Default
    private Instant createdAt = Instant.now();

    @Column(name = "updated_at", nullable = false)
    @Builder.Default
    private Instant updatedAt = Instant.now();

    @PrePersist
    public void prePersist() {
        if (createdAt == null || createdAt.equals(Instant.EPOCH)) {
            createdAt = Instant.now();
        }
        if (updatedAt == null || updatedAt.equals(Instant.EPOCH)) {
            updatedAt = Instant.now();
        }
    }

    @PreUpdate
    public void preUpdate() {
        updatedAt = Instant.now();
    }
}

File: apps/chat/libs/chat-storage/src/main/kotlin/com/example/chat/storage/entity/ChatChannelEntity.kt
package com.example.chat.storage.entity

import com.example.chat.domain.channel.ChannelType
import jakarta.persistence.*
import java.time.Instant

/**
 * 채널 JPA Entity
 */
@Entity
@Table(
    name = "chat_channels",
    indexes = [
        Index(name = "idx_chat_channel_owner", columnList = "owner_id"),
        Index(name = "idx_chat_channel_type", columnList = "channel_type")
    ]
)
class ChatChannelEntity(
    @Id
    @Column(name = "id", length = 36, nullable = false)
    var id: String,

    @Column(name = "name", nullable = false, length = 100)
    var name: String,

    @Column(name = "description", length = 500)
    var description: String? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "channel_type", nullable = false, length = 20)
    var channelType: ChannelType,

    @Column(name = "owner_id", nullable = false, length = 36)
    var ownerId: String,

    @Column(name = "active", nullable = false)
    var active: Boolean = true,

    @Column(name = "created_at", nullable = false)
    var createdAt: Instant = Instant.now(),

    @Column(name = "updated_at", nullable = false)
    var updatedAt: Instant = Instant.now()
) {
    @PrePersist
    fun prePersist() {
        if (createdAt == Instant.EPOCH) {
            createdAt = Instant.now()
        }
        if (updatedAt == Instant.EPOCH) {
            updatedAt = Instant.now()
        }
    }

    @PreUpdate
    fun preUpdate() {
        updatedAt = Instant.now()
    }

    // No-arg constructor for JPA
    protected constructor() : this(
        id = "",
        name = "",
        channelType = ChannelType.PRIVATE,
        ownerId = ""
    )
}

